src/wallet/wallet.cpp

```diff
@@ -1786,23 +1786,37 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate, b
 
     CBlockIndex* pindex = pindexStart;
     {
-        LOCK2(cs_main, cs_wallet);
-
+        LOCK(cs_main);
         // no need to read and scan block, if block was created before
         // our wallet birthday (as adjusted for block time variability)
-        while (pindex && nTimeFirstKey && (pindex->GetBlockTime() < (nTimeFirstKey - 7200)) &&
-                (pindex->nHeight < 1 || !consensus.NetworkUpgradeActive(pindex->nHeight - 1, Consensus::UPGRADE_ZC)))
+        while (pindex && nTimeFirstKey && (pindex->GetBlockTime() < (nTimeFirstKey - 7200)))
             pindex = chainActive.Next(pindex);
+    }
 
+    {
         ShowProgress(_("Rescanning..."), 0); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
-        double dProgressStart = Checkpoints::GuessVerificationProgress(pindex, false);
-        double dProgressTip = Checkpoints::GuessVerificationProgress(chainActive.Tip(), false);
-        std::set<uint256> setAddedToWallet;
+        CBlockIndex* tip = nullptr;
+        double dProgressStart;
+        double dProgressTip;
         std::vector<uint256> myTxHashes;
-        while (pindex) {
-            if (pindex->nHeight % 100 == 0 && dProgressTip - dProgressStart > 0.0)
-                ShowProgress(_("Rescanning..."), std::max(1, std::min(99, (int)((Checkpoints::GuessVerificationProgress(pindex, false) - dProgressStart) / (dProgressTip - dProgressStart) * 100))));
 
+        {
+            LOCK(cs_main);
+            tip = chainActive.Tip();
+            dProgressStart = Checkpoints::GuessVerificationProgress(tip, false);
```

Comment:
good, updated.

Reply:
Actually this line is just redundant, because `dProgressStart` is a constant, so we can set it while initializing it.

We could just remove this code-block by doing
```cpp
        const double dProgressStart = Checkpoints::GuessVerificationProgress(pindex, false);
        const CBlockIndex* tip = WITH_LOCK(cs_main, return chainActive.Tip(); );
        double dProgressTip = Checkpoints::GuessVerificationProgress(tip, false);
```

but imo, there's even a better way: set `tip` and `dProgressTip` at the beginning of the loop (see below), and here init just as
```cpp
        const double dProgressStart = Checkpoints::GuessVerificationProgress(pindex, false);
        const CBlockIndex* tip = nullptr;
        double dProgressTip = 0.0;
```
thus avoiding another cs_main lock.